### PR TITLE
Migrate to ff traits

### PR DESCRIPTION
**Pull request by str4d**
The `Field` trait in this crate is now `FieldExt: ff::PrimeField`.

### CODE REVIEW

File: src/tweedle/fields/fp.rs [hunk 1]
```diff
@@ -569,6 +544,150 @@ impl Field for Fp {
...
+            u64::from_le_bytes(bytes[8..16].try_into().unwrap()),
+            u64::from_le_bytes(bytes[16..24].try_into().unwrap()),
+            u64::from_le_bytes(bytes[24..32].try_into().unwrap()),
+        ];
+
+        BitArray::new(limbs)
+    }
+
+    fn is_odd(&self) -> bool {
+        todo!()
```
Comment (str4d) [inline src/tweedle/fields/fp.rs:615]: Should this be the usual definition?
```suggestion
        self.to_bytes()[0] & 1 == 1
```

File: src/tweedle/fields/fp.rs [hunk 2]
```diff
@@ -569,6 +544,150 @@ impl Field for Fp {
...
+        {
+            BitArray::new(MODULUS_LIMBS_32)
+        }
+
+        #[cfg(target_pointer_width = "64")]
+        BitArray::new(MODULUS.0)
+    }
+
+    fn multiplicative_generator() -> Self {
+        todo!()
```
Comment (str4d) [inline src/tweedle/fields/fp.rs:629]: @ebfull is `GENERATOR` defined anywhere with a different name? Both `ROOT_OF_UNITY` and `DELTA` have doc comments from `GENERATOR`.

File: src/tweedle/fields/fq.rs [hunk 3]
```diff
@@ -584,6 +559,150 @@ impl Field for Fq {
...
+            u64::from_le_bytes(bytes[8..16].try_into().unwrap()),
+            u64::from_le_bytes(bytes[16..24].try_into().unwrap()),
+            u64::from_le_bytes(bytes[24..32].try_into().unwrap()),
+        ];
+
+        BitArray::new(limbs)
+    }
+
+    fn is_odd(&self) -> bool {
+        todo!()
```
Comment (str4d) [inline src/tweedle/fields/fq.rs:630]: Ditto.

File: src/tweedle/fields/fq.rs [hunk 4]
```diff
@@ -584,6 +559,150 @@ impl Field for Fq {
...
+        {
+            BitArray::new(MODULUS_LIMBS_32)
+        }
+
+        #[cfg(target_pointer_width = "64")]
+        BitArray::new(MODULUS.0)
+    }
+
+    fn multiplicative_generator() -> Self {
+        todo!()
```
Comment (str4d) [inline src/tweedle/fields/fq.rs:644]: Ditto.

File: src/tweedle/fields/fq.rs [hunk 5]
```diff
@@ -101,6 +102,19 @@ const MODULUS: Fq = Fq([
...
+#[cfg(not(target_pointer_width = "64"))]
+const MODULUS_LIMBS_32: [u32; 8] = [
+    0x0000_0001,
+    0x842c_afd4,
+    0x6962_86c9,
+    0x038a_a127,
+    0x0000_0000,
+    0x0000_0000,
+    0x0000_0000,
+    0x4000_0000,
```
Comment (daira) [inline src/tweedle/fields/fq.rs:115]: I checked this.

File: src/tweedle/fields/fq.rs [hunk 6]
```diff
@@ -584,6 +559,150 @@ impl Field for Fq {
...
+        0xd023bfdccccccccd,
+        0x360880ec544ed23a,
+        0x3333333333333333,
+        0x3333333333333333,
+    ];
+    const ZETA: Self = Fq::from_raw([
+        0x4394c2bd148fa4fd,
+        0x69cf8de720e52ec1,
+        0x87ad8b5ff9731ffe,
+        0x36c66d3a1e049a58,
```
Comment (daira) [inline src/tweedle/fields/fq.rs:691]: I did *not* check these constants.

**Review by daira — APPROVED** [review]
ACK modulo @str4d's comments.

File: src/tweedle/fields/fp.rs (hunk 1 above)
Comment (ebfull) [reply src/tweedle/fields/fp.rs:615]: Yes.

File: src/tweedle/fields/fp.rs (hunk 2 above)
Comment (ebfull) [reply src/tweedle/fields/fp.rs:629]: `GENERATOR` = 5 for both fields.

Comment (ebfull) [reply src/tweedle/fields/fp.rs:629]: (but it's not defined anywhere in the code currently)

File: src/tweedle/fields/fp.rs [hunk 7]
```diff
@@ -569,6 +553,150 @@ impl Field for Fp {
         CtOption::new(tmp, !self.ct_eq(&Self::zero()))
     }
 
+    fn pow_vartime<S: AsRef<[u64]>>(&self, exp: S) -> Self {
```
Comment (ebfull) [inline src/tweedle/fields/fp.rs:556]: This is provided by `Field` already, right?

File: src/tweedle/fields/fp.rs [hunk 8]
```diff
@@ -176,6 +190,15 @@ const R3: Fp = Fp([
     0x217cb214ebb8fc72,
 ]);
 
+/// `GENERATOR = 5 mod p` is a generator of the `p - 1` order multiplicative subgroup, and
+/// is also a quadratic non-residue.
+const GENERATOR: Fp = Fp([
```
Comment (ebfull) [inline src/tweedle/fields/fp.rs:195]: Can we do `Fp::from_raw([5, 0, 0, 0])` here like we do below?

File: src/tweedle/fields/fq.rs [hunk 9]
```diff
@@ -176,6 +190,15 @@ const R3: Fq = Fq([
     0xa42cdf95f959577,
 ]);
 
+/// `GENERATOR = 5 mod q` is a generator of the `q - 1` order multiplicative subgroup, and
+/// is also a quadratic non-residue.
+const GENERATOR: Fq = Fq([
```
Comment (ebfull) [inline src/tweedle/fields/fq.rs:195]: Similarly, `Fq::from_raw([5, 0, 0, 0])`

File: src/tweedle/fields/fq.rs [hunk 10]
```diff
@@ -584,6 +568,150 @@ impl Field for Fq {
         CtOption::new(tmp, !self.ct_eq(&Self::zero()))
     }
 
+    fn pow_vartime<S: AsRef<[u64]>>(&self, exp: S) -> Self {
```
Comment (ebfull) [inline src/tweedle/fields/fq.rs:571]: Similarly, this is provided by `Field` already.

**Review by ebfull — CHANGES_REQUESTED** [review]

File: src/tweedle/fields/fp.rs (hunk 7 above)
Comment (str4d) [reply src/tweedle/fields/fp.rs:556]: Yes, but the `ff` implementation doesn't skip squarings before finding a one, and the existing impl here did, so I assumed you wanted this behaviour.

**Review by ebfull — APPROVED** [review]

**Pull request merged**